scipy/optimize/_differentialevolution.py

```diff
@@ -598,6 +619,43 @@ def __init__(self, func, bounds, args=(),
 
         self.random_number_generator = check_random_state(seed)
 
+        # Which parameters are going to be integers?
+        if np.any(integrality):
+            # user has provided a truth value for integer constraints
+            if polish:
+                message = ("Polishing is incompatible with integrality"
+                           " constraints; `polish` will be ignored.")
+                self.polish = False
+                warnings.warn(message, OptimizeWarning)
+
+            self.integrality = np.broadcast_to(
+                integrality,
+                self.parameter_count
+            )
+            self.integrality = np.asarray(self.integrality, bool)
+            # For integrality parameters change the limits to only allow
+            # integer values lying between the limits.
+            lb, ub = np.copy(self.limits)
+
+            lb[lb > 0] = np.ceil(lb[lb > 0])
+            lb[lb < 0] = np.trunc(lb[lb < 0])
+            ub[ub > 0] = np.trunc(ub[ub > 0])
+            ub[ub < 0] = np.floor(ub[ub < 0])
+            if not (lb[integrality] <= ub[integrality]).all():
```

Comment:
This needs to be strict inequality. There needs to be an integer between them.

Reply:
That's incorrect (this is before 0.5 is added/subtracted). e.g. an original bounds of (0.9, 1.1) is transformed to [1, 1]. There is therefore one valid value (1) it can take. Also, it's allowed in `differential_evolution` to have `lb == ub`, so a valid original bound (for an integrality constraint or otherwise) could also be `(1, 1)`.

Reply:
>  but (0.9, 0.9) will have no solutions if it's integer constrained.

When `ceil` and `trunc` are applied the bounds will transform to (1.0, 0.0), which will fail raise the `ValueError` because `ub` isn't less than or equal to `lb`.

Reply:
Of course. Good call.